crates/trippy-tui/src/locale.rs

```diff
@@ -1,43 +1,125 @@
+    };
+    set_locale_inner(new_locale)
 }
 
-/// Get the current locale.
-pub fn locale() -> String {
-    rust_i18n::locale().to_string()
+/// Get the available locales.
+pub fn available_locales() -> Vec<&'static str> {
+    data()
+        .0
+        .iter()
+        .flat_map(|(_, v)| v.0.keys().map(AsRef::as_ref))
+        .unique()
+        .sorted_unstable()
+        .collect::<Vec<_>>()
 }
 
-/// Get all available locales.
-pub fn available_locales() -> Vec<&'static str> {
-    rust_i18n::available_locales!()
+/// A macro to translate an item to the current locale.
+#[macro_export]
+macro_rules! t {
+    ($key:expr) => {
+        std::borrow::Cow::Borrowed($crate::locale::__translate($key))
+    };
+    ($key:expr, $($kt:ident = $kv:expr),+) => {
+        {
+            let string = t!($key);
+            $(
+                let string = string.replace(concat!("%{", stringify!($kt), "}"), &$kv.to_string());
+            )+
+            string
+        }
+    };
+    ($key:expr, $($kt:literal => $kv:expr),+) => {
+        {
+            let string = t!($key);
+            $(
+                let string = string.replace(concat!("%{", $kt, "}"), &$kv.to_string());
+            )+
+            string
+        }
+    };
+}
+
+/// Translate an item to the current locale.
+///
+/// This function is public as it is used by the `t!` macro, however is not considered part of the
+/// public interface.
+#[doc(hidden)]
+pub fn __translate(item: &str) -> &str {
+    translate_locale(item, locale().load().as_str())
+}
+
+/// Translate an item to a specific locale.
+///
+/// If the item does not exists, the key is returned. Otherwise, if item does not contain the
+/// locale, the fallback locale is used. If the fallback locale does not exist, the key is
+/// returned.
+fn translate_locale<'a>(item: &'a str, locale: &'_ str) -> &'a str {
+    if let Some(key) = data().0.get(item) {
+        if let Some(value) = key.0.get(locale) {
+            value
+        } else if let Some(value) = key.0.get(&split_locale(locale)) {
+            value
+        } else if let Some(value) = key.0.get(FALLBACK_LOCALE) {
+            value
+        } else {
+            item
+        }
+    } else {
+        item
+    }
+}
+
+/// Get the locale data.
+fn data() -> &'static Data {
+    static DATA: OnceLock<Data> = OnceLock::new();
+    DATA.get_or_init(|| {
+        toml::from_str(include_str!("../locales.toml")).expect("Failed to parse locales.toml")
+    })
 }
 
-fn set_locale_inner(locale: &str) {
-    let all_locales = rust_i18n::available_locales!();
-    if all_locales.contains(&locale) {
-        rust_i18n::set_locale(locale);
+/// This is a map of a item name (i.e. `title_hops`, `awaiting_data`, etc.) to the locale `Item`.
+#[derive(Debug, Deserialize)]
+struct Data(HashMap<String, Item>);
+
+/// This is a map of locale keys (i.e. `en`, `zh`, etc.) to the translated value.
+#[derive(Debug, Deserialize)]
+struct Item(HashMap<String, String>);
```

Comment:
The key here (the locale string) could be a concrete type i.e. (`Locale`) with validation rules etc.  Overkill for now I'd say.